### PR TITLE
Add support for DirectResponse in agw

### PR DESCRIPTION
**Pull request by corsairier**
# Description
Closes: #11843 

# Change Type

Select one or more of the following by including the corresponding slash-command:
```
/kind new_feature
```

# Changelog

```release-note
Added DirectResponse Support in agentgateway
```

# Additional Notes

Tested with AgentGateway
1. Setup the kind cluster and build images using this command `./hack/kind/setup-kind.sh`
2. Install CRDs and kgateway control plane using 

```
helm upgrade -i -n kgateway-system kgateway-crds _test/kgateway-crds-1.0.0-ci1.tgz --version 1.0.0-ci1 \
          --create-namespace

helm upgrade -i -n kgateway-system kgateway _test/kgateway-1.0.0-ci1.tgz --version 1.0.0-ci1 \
          --set image.registry=ghcr.io/kgateway-dev \
          --set agentGateway.enabled=true \
          --create-namespace
```
4. Applying DirectResponse config

```
kubectl apply -f- <<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: agent-gateway
  namespace: kgateway-system
spec:
  gatewayClassName: agentgateway
  listeners:
  - name: http
    protocol: HTTP
    port: 80
    allowedRoutes:
      namespaces:
        from: All
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: direct-response-route
  namespace: kgateway-system
spec:
  parentRefs:
  - name: agent-gateway
    namespace: kgateway-system
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /direct
    filters:
    - type: ExtensionRef
      extensionRef:
        group: gateway.kgateway.dev
        kind: DirectResponse
        name: test-direct-response
---
apiVersion: gateway.kgateway.dev/v1alpha1
kind: DirectResponse
metadata:
  name: test-direct-response
  namespace: kgateway-system
spec:
  status: 200
  body: "Direct response from agent gateway"
EOF
```

4. Port Forwarding: 
```
kubectl port-forward -n kgateway-system deployment/agent-gateway 8080:80
```

5. Send request:

```
> curl -v localhost:8080/direct
* Host localhost:8080 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8080...
* Connected to localhost (::1) port 8080
> GET /direct HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/8.5.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< content-length: 34
< date: Thu, 07 Aug 2025 09:53:23 GMT
< 
* Connection #0 to host localhost left intact
Direct response from agent gateway
```

### CODE REVIEW

**Comment by corsairier** [comment]
Hi @npolshakova,  
I’ve opened this draft PR and wanted to get some early feedback. I haven’t written E2E tests or other related tests yet, and I’m not sure if there are additional steps or requirements I might be missing. Could you please review my changes and let me know:

- If the implementation and approach are correct
- What tests or documentation should be added
- Any other improvements or requirements before this can move out of draft

Thank you for your guidance!

**Comment by npolshakova** [comment]
Hi @corsairier! 👋  Thanks for the contribution! 

I would add the following tests:
- unit test for the filter logic: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/agentgatewaysyncer/route_collections_test.go#L1552
- translator-level test for the syncer: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/agentgatewaysyncer/agentgateway_translator_test.go 
- env test: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/setup/agentgateway_test.go 

With that coverage we probably don't need a dedicated e2e test, but I would write the tests you manually took to validate in the PR description so we can have an example for future docs! 

We'll also want to consider the case when there's already a route action set, similar how the envoy plugin does it: https://github.com/kgateway-dev/kgateway/blob/911f7a02c4fee803b535bd53152de851e00b00e2/internal/kgateway/extensions2/plugins/directresponse/direct_response_plugin.go#L121

**Comment by corsairier** [comment]
Thanks for your feedback, @npolshakova! I'll make those changes!

**Comment by corsairier** [comment]
@npolshakova, I have added the following tests,


> * unit test for the filter logic: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/agentgatewaysyncer/route_collections_test.go#L1552
> * translator-level test for the syncer: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/agentgatewaysyncer/agentgateway_translator_test.go

Also I have added the logic for this edge case,
> We'll also want to consider the case when there's already a route action set, similar how the envoy plugin does it:
> 
> https://github.com/kgateway-dev/kgateway/blob/911f7a02c4fee803b535bd53152de851e00b00e2/internal/kgateway/extensions2/plugins/directresponse/direct_response_plugin.go#L121

However, I am not sure about how to implement the env test. Could you provide some guidance or examples on how to approach writing env test here?

**Comment by npolshakova** [comment]
Yep, the agentgateway env tests are run here: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/setup/agentgateway_test.go 

The test files live here where you can find some samples: https://github.com/kgateway-dev/kgateway/tree/main/internal/kgateway/setup/testdata/agentgateway. These inputs are the same as the translator-level testdata, but the env tests check that the xds is being correctly served.

Once you create an input file, if you run the test it should output a file with the config dump that you can check looks as expected with the direct response route.

**Comment by corsairier** [comment]
Hi @npolshakova,
Thanks for your guidance. This PR is now ready for review. I have added the env tests along with the tests I manually took in the PR description.

File: internal/kgateway/agentgatewaysyncer/conversion.go [hunk 1]
```diff
@@ -110,6 +108,10 @@ func convertHTTPRouteToADP(ctx RouteContext, r gwv1.HTTPRouteRule,
 	res.Hostnames = slices.Map(obj.Spec.Hostnames, func(e gwv1.Hostname) string {
 		return string(e)
 	})
+	// Return filter error if present, otherwise return backend error
+	if filterError != nil {
+		return res, filterError
+	}
 	return res, backendErr
```
Comment (npolshakova) [inline internal/kgateway/agentgatewaysyncer/conversion.go]: Let's do a multi error here (create a list of errs, append and then errors.Join(errs...)

File: internal/kgateway/agentgatewaysyncer/agentgateway_translator_test.go [hunk 2]
```diff
@@ -192,4 +192,100 @@ var _ = DescribeTable("Basic agentgateway Tests",
...
+			Expect(resolvedRefs.Status).To(Equal(metav1.ConditionTrue))
+			Expect(resolvedRefs.Reason).To(Equal(string(gwv1.RouteReasonResolvedRefs)))
+
+			// Assert Accepted=False reports the missing DirectResponse
+			acceptedCond := meta.FindStatusCondition(routeStatus.Parents[0].Conditions, string(gwv1.RouteConditionAccepted))
+			Expect(acceptedCond).NotTo(BeNil())
+			Expect(acceptedCond.Status).To(Equal(metav1.ConditionFalse))
+			Expect(acceptedCond.Reason).To(Equal(string(gwv1.RouteReasonBackendNotFound)))
+			Expect(acceptedCond.Message).To(ContainSubstring("DirectResponse"))
+			Expect(acceptedCond.Message).To(ContainSubstring("not found"))
```
Comment (npolshakova) [inline internal/kgateway/agentgatewaysyncer/agentgateway_translator_test.go]: Let's assert on the full error message instead of checking the message for different substrings.

File: internal/kgateway/agentgatewaysyncer/conversion.go [hunk 3]
```diff
@@ -301,15 +305,27 @@ func buildADPFilters(
 			}
 			filters = append(filters, h)
 		case gwv1.HTTPRouteFilterRequestRedirect:
+			if hasTerminalFilter {
+				filterError = &reporter.RouteCondition{
+					Type:    gwv1.RouteConditionAccepted,
+					Status:  metav1.ConditionFalse,
+					Reason:  gwv1.RouteReasonIncompatibleFilters,
+					Message: "Cannot combine RequestRedirect with another terminal filter (e.g., DirectResponse).",
```
Comment (npolshakova) [inline internal/kgateway/agentgatewaysyncer/conversion.go]: Let's define this common error at the top level:
```
func terminalFilterCombinationError(existingFilter, newFilter string) string {
    return fmt.Sprintf("Cannot combine multiple terminal filters: %s and %s are mutually exclusive. Only one terminal filter is allowed per route rule.", existingFilter, newFilter)
}
```

**Comment by npolshakova** [comment]
@corsairier looks like there are some merge conflicts. Let me know when they're resolved and I can re-approve!

File: internal/kgateway/setup/testdata/agentgateway/grpcroute-out.json [hunk 4]
```diff
@@ -22,6 +22,27 @@
         }
       }
     },
+    {
```
Comment (npolshakova) [inline internal/kgateway/setup/testdata/agentgateway/grpcroute-out.json:25]: Hm, it's a little strange that the direct response service is getting added to the grpc, httproute, tcproute and inferencepool tests when those aren't using the config. It might just be stale config in the env test. 

Let's double check the config get cleaned up after applying the example, then deleting the direct response:
```
kubectl apply -f- <<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: agent-gateway
  namespace: kgateway-system
spec:
  gatewayClassName: agentgateway
  listeners:
  - name: http
    protocol: HTTP
    port: 80
    allowedRoutes:
      namespaces:
        from: All
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: direct-response-route
  namespace: kgateway-system
spec:
  parentRefs:
  - name: agent-gateway
    namespace: kgateway-system
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /direct
    filters:
    - type: ExtensionRef
      extensionRef:
        group: gateway.kgateway.dev
        kind: DirectResponse
        name: test-direct-response
---
apiVersion: gateway.kgateway.dev/v1alpha1
kind: DirectResponse
metadata:
  name: test-direct-response
  namespace: kgateway-system
spec:
  status: 200
  body: "Direct response from agent gateway"
EOF
```

Comment (corsairier) [reply internal/kgateway/setup/testdata/agentgateway/grpcroute-out.json:25]: Would you recommend running the tests individually using TEST_PREFIX? This removed the direct response service for me

Comment (npolshakova) [reply internal/kgateway/setup/testdata/agentgateway/grpcroute-out.json:25]: Okay perfect, that means there's a caching issue between the test runs. I'll create an issue to look into why it's not getting cleared between tests, but that can be done separately.

Comment (npolshakova) [reply internal/kgateway/setup/testdata/agentgateway/grpcroute-out.json:25]: https://github.com/kgateway-dev/kgateway/issues/11968

**Comment by npolshakova** [comment]
@corsairier looks like there's a merge conflict. Let me know when it's resolved and I can re-approve.

**Comment by corsairier** [comment]
@npolshakova, Now its resolved

**Comment by npolshakova** [comment]
@corsairier looks like there's a small lint error. You can test this locally by running `make verify`. https://github.com/kgateway-dev/kgateway/actions/runs/16913232984/job/47928345509?pr=11859#step:4:178

Running `make generate-all` should also fix the formatting for the future.
